### PR TITLE
Make CGroups CPU period configurable

### LINKED ISSUE #1627: Make CFS quota period configurable
It seems that the CFS quota period is fixed in ecs agent for the parent cgroup that is created for each task. https://github.com/aws/amazon-ecs-agent/blob/d68e729f73e588982dc2189a1c618c18c47c931b/agent/api/task/task_linux.go#L39

Since the length of this period can have consequences on performance (see [this](https://github.com/kubernetes/kubernetes/issues/51135) and [this](https://gist.github.com/bobrik/2030ff040fad360327a5fab7a09c4ff1)) we would like it to be configurable, e.g. as an argument to ecs-agent.

### PR DESCRIPTION
**Pull request by boynux**
Signed-off-by: Mohamad Arab <boynux@gmail.com>

### Summary
Fixes #1627
Reopen: #1632 
Make CFS CPU Period configurable

CFS keeps throttling tasks very excessively if the default period 100ms is specified, to reduce the impact we should be able to set lower CPU period when defining cgroup quotas.

### Implementation details
New environment parameter added to adjust CFS period.

### Testing
See here: https://gist.github.com/bobrik/2030ff040fad360327a5fab7a09c4ff1

New tests cover the changes: yes

### Description for the changelog

Configure CFS scheduler evaluation period from 8ms to 100ms

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by boynux** [comment]
@adnxn I applied the suggested changes here.

**Comment by boynux** [comment]
@adnxn looks like some tests are failing, can I get more details to see if they are related to this change. I remembered last time they all passed.

**Comment by adnxn** [comment]
> @adnxn looks like some tests are failing, can I get more details to see if they are related to this change. I remembered last time they all passed.

ill have to dig into this and try to get you some more information. in the meantime - have you looked at running the make targets for the integ and functional tests?

**Comment by adnxn** [comment]
@boynux: do you mind rebasing your changes on the latest changes from `dev`? just noticed that the version file in your branch is pointing to `1.25.3` and you're missing some fixes we've added to testing since then. the three functional tests that appear to be failing all have been fixed in dev branch.

**Comment by boynux** [comment]
@adnxn just rebased the latest dev and push the update. Let's see

**Comment by boynux** [comment]
@adnxn is there anything I can do for this change?

**Comment by boynux** [comment]
@adnxn sorry to  bother you again, is this still in your radar? And is there anything I can do to help getting this rolled out?

**Comment by adnxn** [comment]
> @adnxn sorry to bother you again, is this still in your radar? And is there anything I can do to help getting this rolled out?

no worries, yea it fell off my radar again. we'll have a second person go over the code as well and hopefully get this wrapped up soon. thanks for checking in again.

File: agent/api/task/task.go [hunk 1]
```diff
@@ -250,7 +250,7 @@ func (task *Task) PostUnmarshalTask(cfg *config.Config,
 	// hook into this
 	task.adjustForPlatform(cfg)
 	if task.MemoryCPULimitsEnabled {
-		err := task.initializeCgroupResourceSpec(cfg.CgroupPath, resourceFields)
+		err := task.initializeCgroupResourceSpec(cfg.CgroupPath, cfg.CGroupCPUPeriod, resourceFields)
```
Comment (fenxiong) [inline agent/api/task/task.go]: to follow naming convention, can you change "CGroup" to "Cgroup"? same for other places

File: agent/config/config.go [hunk 2]
```diff
@@ -133,6 +133,11 @@ const (
 
 	// DefaultNvidiaRuntime is the name of the runtime to pass Nvidia GPUs to containers
 	DefaultNvidiaRuntime = "nvidia"
+
+	// DefaultCPUPeriod is set to 100 ms to set CFS period and quota for task limits
```
Comment (fenxiong) [inline agent/config/config.go]: DefaultCPUPeriod -> defaultCgroupCPUPeriod

**Comment by boynux** [comment]
@fenxiong done as suggested

**Comment by adnxn** [comment]
@boynux - just to update you im trying to get these changes tested through our test suites and merged.

in the meantime - since this doesnt have any functional tests. would you be able to test and report the failure modes of full task runs at the boundaries of the configurable period?

**Comment by adnxn** [comment]
okay so looks like we're getting `ecs/windows/integ_test — Failed: Windows integration test suite
` cause your branch is missing [this commit](https://github.com/aws/amazon-ecs-agent/commit/9baa7702fca2125a341cec34bdc8c7f7432d7f70). mind rebasing off dev once more? 😀

**Comment by yunhee-l** [comment]
integ test had failed due to flakey test (TestShutdownOrder): https://github.com/aws/amazon-ecs-agent/issues/1931. Test was rerun and all passed now.

**Comment by boynux** [comment]
Thanks for approving this change @adnxn

**Comment by boynux** [comment]
🎉

**Pull request merged**